src/arrays/spinops.jl

```diff
@@ -0,0 +1,110 @@
+#########################
+# Spin Operator Methods #
+#########################
+#########################
+# Implementation  -  1  #
+#########################
+#function spin_h1(j::Float64)
+#    if (isinteger(2*j) && j>0)
+#        m = reverse([-j:j-1])
+#        N = length(m)+1
+#        m_coeffs = [sqrt(j*(j+1.0) - (x+1.0)*x) for x in m]
+#        return spdiagm(m_coeffs,1,N,N)
+#    else
+#        return error("j must be non-negative integer or half integer")
+#  end
+#end
+
+#function spin_h2(j::Float64)
+#    if (isinteger(2*j) && j>0)
+#        m = reverse([-j:j])
+#        N = length(m)
+#        return spdiagm(m,0,N,N)
+#    else
+#        return error("j must be non-negative integer or half integer")
+#  end
+#end
+
+#spin_jx(j::Float64) = QuArray(0.5*(spin_h1(j)+spin_h1(j)'))
+#spin_jy(j::Float64) = QuArray(-0.5*im*(spin_h1(j)-spin_h1(j)'))
+#spin_jz(j::Float64) = QuArray(spin_h2(j))
+#spin_jp(j::Float64) = QuArray(spin_h1(j))
+#spin_jm(j::Float64) = QuArray(spin_h1(j)')
+#const sigma_x = 2.0 * spin_jx(0.5)
+#const sigma_y = 2.0 * spin_jy(0.5)
+#const sigma_z = 2.0 * spin_jz(0.5)
+
+# TODO : unicode sigma_y, sigma_z
+#const σₓ = sigma_x
+
+#########################
+# Implementation  -  2  #
+#########################
+#abstract Spin{S}
+
+immutable HalfSpin{S} end
+
+# constructors
+HalfSpin(j::Integer) = j > 0 ? HalfSpin{2*j}() : error("Spin must be positive")
+
+function HalfSpin(j::Float64)
+    if (j > 0) && isinteger(j)
+        HalfSpin(int(j))
+    elseif (j > 0) && isinteger(j*2)
+        HalfSpin{int(2*j)}()
+    else
+        error("Spin must be positive and a multiple of 1/2.")
+    end
+end
+
```

Comment:
In the above, `j` moves from the value domain to the type domain, making this a type-unstable constructor. In addition, we multiply `j` by 2, which I find confusing for this constructor, since `HalfSpin{S}()` represents a spin value of `S * 1/2`. Thus, if I'm passing in `3` to a constructor for `HalfSpin`, I would expect that to return a representation of the spin value `3 *1/2`, rather than a total spin of `3`. I think this kind of doubling would be better to implement in the `spin` function below instead.

Also, the checks being done here become pointless, since I would bypass it just by calling `HalfSpin{2}()` instead of calling `HalfSpin(2)` (the former is also more natural to me anyway, since it's type-stable).

I say we should just remove the above constructors for `HalfSpin`, and just re-define the type like this:

```
immutable HalfSpin{S} 
    function HalfSpin()
        S > 0 ? new() : error("Spin must be positive")
    end
end
```

...which provides a single, safe, type-stable constructor.

Then, we can let the `spin` function below handle converting a numeric argument to the appropriate spin value (with additional checks, doubling, etc.). The `spin` function would still be type-unstable, since eventually the argument would pass from the value domain to the type domain, but I don't think it will matter that much as long as that function isn't directly used for heavy computations.